CodeEdit/Features/LSP/LanguageServer/LSPContentCoordinator.swift

```diff
@@ -0,0 +1,92 @@
+//
+//  LSPContentCoordinator.swift
+//  CodeEdit
+//
+//  Created by Khan Winter on 9/12/24.
+//
+
+import AppKit
+import AsyncAlgorithms
+import CodeEditSourceEditor
+import CodeEditTextView
+import LanguageServerProtocol
+
+/// This content coordinator forwards content notifications from the editor's text storage to a language service.
+///
+/// This is a text view coordinator so that it can be installed on an open editor. It is kept as a property on
+/// ``CodeFileDocument`` since the language server does all it's document management using instances of that type.
+///
+/// Language servers expect edits to be sent in chunks (and it helps reduce processing overhead). To do this, this class
+/// keeps an async stream around for the duration of its lifetime. The stream is sent edit notifications, which are then
+/// chunked into 250ms timed groups before being sent to the ``LanguageServer``.
+class LSPContentCoordinator: TextViewCoordinator, TextViewDelegate {
+    // Required to avoid a large_tuple lint error
+    private struct SequenceElement: Sendable {
+        let uri: String
+        let range: LSPRange
+        let string: String
+    }
+
+    private var editedRange: LSPRange?
+    private var stream: AsyncStream<SequenceElement>?
+    private var sequenceContinuation: AsyncStream<SequenceElement>.Continuation?
+    private var task: Task<Void, Never>?
+
+    weak var languageServer: LanguageServer?
+    var uri: String?
+
+    init() {
+        self.stream = AsyncStream { continuation in
+            self.sequenceContinuation = continuation
+        }
+    }
+
+    func setUpUpdatesTask() {
+        task?.cancel()
+        guard let stream else { return }
+        task = Task.detached { [weak self] in
+            // Send edit events every 250ms
+            for await events in stream.chunked(by: .repeating(every: .milliseconds(250), clock: .continuous)) {
+                guard !Task.isCancelled, self != nil else { return }
+                guard !events.isEmpty, let uri = events.first?.uri else { continue }
+                // Errors thrown here are already logged, not much else to do right now.
+                try? await self?.languageServer?.documentChanged(
+                    uri: uri,
+                    changes: events.map {
+                        LanguageServer.DocumentChange(replacingContentsIn: $0.range, with: $0.string)
+                    }
+                )
+            }
+        }
+    }
+
+    func prepareCoordinator(controller: TextViewController) {
```

Comment:
I think the input parameter is unnecessary
```suggestion
    func prepareCoordinator() {
```

Reply:
This parameter is a requirement of the [text coordinator protocol](https://codeeditapp.github.io/CodeEditSourceEditor/documentation/codeeditsourceeditor/textviewcoordinator/preparecoordinator(controller:)) so it can't be removed.

Reply:
Oh sorry 🤦‍♂️